agentendpoint/exec_task.go

```diff
@@ -0,0 +1,106 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package agentendpoint
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"time"
+
+	"github.com/GoogleCloudPlatform/osconfig/config"
+
+	agentendpointpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/agentendpoint/v1alpha1"
+)
+
+var (
+	winRoot = os.Getenv("SystemRoot")
+	sh      = "/bin/sh"
+
+	winPowershell string
+	winCmd        string
+
+	goos = runtime.GOOS
+
+	errLinuxPowerShell = errors.New("interpreter POWERSHELL cannot be used on non-Windows system")
+	errWinNoInt        = fmt.Errorf("interpreter must be specified for a Windows system")
+)
+
+type execTask struct {
+	client *Client
+
+	TaskID    string
+	Task      *execStepTask
+	StartedAt time.Time         `json:",omitempty"`
+	LogLabels map[string]string `json:",omitempty"`
+}
+
+type execStepTask struct {
+	*agentendpointpb.ExecStepTask
+}
+
+func (e *execTask) reportCompletedState(ctx context.Context, errMsg string, output *agentendpointpb.ReportTaskCompleteRequest_ExecStepTaskOutput) error {
+	req := &agentendpointpb.ReportTaskCompleteRequest{
+		TaskId:       e.TaskID,
+		TaskType:     agentendpointpb.TaskType_EXEC_STEP_TASK,
+		ErrorMessage: errMsg,
+		Output:       output,
+	}
+	if err := e.client.ReportTaskComplete(ctx, req); err != nil {
+		return fmt.Errorf("error reporting completed state: %v", err)
+	}
+	return nil
+}
+
+func (e *execTask) run(ctx context.Context) error {
```

Comment:
What happens when the agent reportTaskProgress and then the instance is restarted. Would the agent not pick up the exec step again?

Reply:
Exec doesnt have state we wish to save, we don't allow for recovery from a crash or reboot. We are leaving the logic up to the server though, so if we reportTaskProgress then crash and restart I would expect the server to record it as a failure as we lost our place for some unknown reason.

Reply:
Before starting the execution and after calling ReportTaskProgress, if the instance restarts - the server expects that the agent will finish the execution, so we should be saving the task to execute. While the agent was running the task, if the instance restarts then yes that should be marked as failure. What do you think?

Reply:
There is no reason the agent should restart at that point, or any reason there will be a gap of any serious amount of time between ReportTaskProgress and the actual execution (we're talking ms here, it reports started, then checks if it should continue, then starts execution). Since we don't know why the agent restarted or where we may be in the execution process and we do not know how to recover so we should just treat it as a failure (the server will, the agent will be in an odd state). We could save state so the agent could come back up and know it is in a failed state and then send the failure report, but that's adding additional logic to the agent that probably doesn't need to be there (unless you tell me that will be simpler than the service doing it).

See this PR for the rest of the exec logic: https://github.com/GoogleCloudPlatform/osconfig/pull/86/